### PR TITLE
Adding Conformance Test for the Java Firestore SDK

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds the conformance test suite to the Firestore SDK.

Functional changes includes new validation:
- DocumentMasks are now ordered.
- Deletes are only allowed for update() and set() with merge:true
- Preconditions are only added when non-empty
- Field masks are validated to contain only elements listed in the input data.
- Data for update() and set() with merge cannot be empty.
- ~~Query orderBy constraint have to be unique per field path~~ (no longer true after Watch merge)
- Precondition.exists() can no longer be specified manually on update() calls.

This PR also moves field encoding logic from FirestoreImpl to UserDataConverter.

### CODE REVIEW

File: google-cloud-firestore/src/main/java/com/google/cloud/firestore/Query.java [hunk 1]
```diff
@@ -41,7 +41,10 @@
 import com.google.protobuf.Int32Value;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply google-cloud-firestore/src/main/java/com/google/cloud/firestore/Query.java]: This breaks the consistency between the instance variables. I'd rather leave as is.

**Comment by schmidt-sebastian** [comment]
@garrettjonesgoogle / @pongad: Am I ok to commit this change? Thanks!

**Comment by pongad** [comment]
LGTM. @garrettjonesgoogle PTAL

**Comment by garrettjonesgoogle** [comment]
I will defer to @pongad for this one.

**Pull request merged**